proposals/2024-06-01-Self-Identification-and-Bundling.md

```diff
@@ -0,0 +1,123 @@
+
+As [OAS 3.1.1 clarifies](https://github.com/OAI/OpenAPI-Specification/pull/3758), it is already mandatory to separate location and identity for Schema Object support.  Currently, associating a URI other than the current URL with a document to meet this requirement has to be done externally.  Many tools effectively support this in the form of allowing the retrieval URL to be set manually, without verifying that the document actually lives at the given URL.
+
+The various streaming formats do not state how to resolve links among the parts, as noted in [RFC 9512 YAML Media Type §3.2](https://www.rfc-editor.org/rfc/rfc9512.html#name-yaml-streams), which makes an explicit analogy to `application/json-seq` for this behavior.  
+
+A `self` field that is a relative URI-reference would be resolved against the document location just as all Reference Object and similar URIs are resolved in OAS 3.1.  A relative reference within a stream would resolve against the URL of the entire stream; however it is probably better to either RECOMMEND or require (MUST) resolving `self` to an absolute URI when bundling into a stream for maximum predictability.
+
+## Detailed design
+
+This is written for the structure of the OAS, but it should be clear how it would be adapted for Arazzo.
+
+\# Specification
+
+...
+
+\#\# OpenAPI Description Structure
+
+...
+
+\#\#\# Bundling Documents as YAML or JSON Streams
+
+Multiple OpenAPI Description documents MAY be bundled in a YAML stream ([RFC 9512 §3.2](https://www.rfc-editor.org/rfc/rfc9512.html#name-yaml-streams)) or a JSON streaming format such as JSON Text Sequences [RFC 7464](https://www.rfc-editor.org/rfc/rfc7464), [JSON Lines](https://jsonlines.org/), or [NDJSON](https://github.com/ndjson/ndjson-spec).  Documents bundled in this way MUST set their own URI using the `self` field in the [OpenAPI Object](#oasObject).  If a document in the stream has `self` set to a relative URI-reference, it MUST be resolved relative to the location of the entire stream.  However, it is strongly RECOMMENDED to set `self` to an absolute URI for use within multi-document streams.
+
+The first document in the stream MUST be treated as the entry document.
+
+\# Schema
+
+...
+
+\#\# OpenAPI Object
+
+\#\#\# Fixed Fields
+
+Field Name | Type | Description
+---|:---|:---
+self | `URI-reference` (without a fragment) | Sets the URI of this document, which also serves as its base URI in accordance with [RFC 3986 §5.1.1](https://www.rfc-editor.org/rfc/rfc3986#section-5.1.1); the value MUST NOT be the empty string and MUST NOT contain a fragment
```

Comment:
For interleaving multiple OAD streams, we could also include an optional field to identify the bundle, perhaps using the entry document ID.  Ideally, it would allow multiple bundles to prevent duplicating documents in certain scenarios, but I'm usually the odd one out for being cool with arrays.

Reply:
@kevinswiber I think this would be a good further direction, but I worry about making the initial thing too complicated.  Really, the only thing you need to solve for having multiple OADs in a stream is how to identify each entry document.  This proposal just says "the first document in the stream is the entry document", which means there can only be one clear OAD in the stream.  But there's no reason you couldn't get a stream, and then select (by URI) each entry document you want to try to parse from the stream.  As long as all needed documents are in the stream, it would work fine.